snuba/datasets/metrics_aggregate_processor.py

```diff
@@ -0,0 +1,137 @@
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Mapping, Optional
+
+from snuba.clickhouse.value_types import (
+    ClickhouseInt,
+    ClickhouseNumArray,
+    ClickhouseUnguardedExpression,
+)
+from snuba.consumers.types import KafkaMessageMetadata
+from snuba.processor import (
+    AggregateInsertBatch,
+    MessageProcessor,
+    ProcessedMessage,
+    _ensure_valid_date,
+)
+
+
+class MetricsAggregateProcessor(MessageProcessor, ABC):
+    GRANULARITIES = [10, 3600, 3600 * 24]
+
+    @abstractmethod
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    def process_message(
+        self, message: Mapping[str, Any], metadata: KafkaMessageMetadata
+    ) -> Optional[ProcessedMessage]:
+        # TODO: Support messages with multiple buckets
+
+        if not self._should_process(message):
+            return None
+
+        timestamp = _ensure_valid_date(datetime.utcfromtimestamp(message["timestamp"]))
+        assert timestamp is not None
+
+        keys = []
+        values = []
+        tags = message["tags"]
+        assert isinstance(tags, Mapping)
+        for key, value in sorted(tags.items()):
+            assert key.isdigit()
+            keys.append(int(key))
+            assert isinstance(value, int)
+            values.append(value)
+
+        processed = [
+            {
+                "org_id": ClickhouseInt(message["org_id"]),
+                "project_id": ClickhouseInt(message["project_id"]),
+                "metric_id": ClickhouseInt(message["metric_id"]),
+                "timestamp": ClickhouseUnguardedExpression(
+                    f"toStartOfInterval(toDateTime('{timestamp.isoformat()}'), toIntervalSecond({granularity}))"
+                ),
+                "tags.key": ClickhouseNumArray(keys),
+                "tags.value": ClickhouseNumArray(values),
+                **self._process_values(message),
+                "retention_days": ClickhouseInt(message["retention_days"]),
+                "partition": metadata.partition,
+                "offset": metadata.offset,
+                "granularity": ClickhouseInt(granularity),
+            }
+            for granularity in self.GRANULARITIES
+        ]
+        return AggregateInsertBatch(processed, None)
+
+
+class SetsAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return message["type"] is not None and message["type"] == "s"
```

Comment:
These message types look really arbitrary. "s" for sets. "c" for counters. "d" for distributions

Can you define all the metrics message types somewhere and refer to them in the processing logic.

Also, what was the reason to not have the message type be "set" or "counters"?

Reply:
This is something I've inherited. @MeredithAnya do you know if we documented the rationale for this somewhere?

I'm down to extract these magic constants out, though

Reply:
@onewland @nikhars yeah this was set in relay I believe, I also inherited this. @jan-auer maybe you have context on the rationale?